salt/pillar/__init__.py

```diff
@@ -145,10 +145,10 @@ def __init__(self, opts, grains, minion_id, saltenv, ext=None, functions=None,
         self.channel = salt.transport.client.AsyncReqChannel.factory(opts)
         if pillarenv is not None:
             self.opts['pillarenv'] = pillarenv
-        self.pillar_override = pillar_override or {}
-        if not isinstance(self.pillar_override, dict):
-            self.pillar_override = {}
-            log.error('Pillar data must be a dictionary')
+        self.pillar_override = pillar_override or OrderedDict()
+        if not isinstance(self.pillar_override, dict) and not isinstance(self.pillar_override, OrderedDict):
```

Comment:
OrderedDict inherits from `dict`, the second isinstance check is redundant:

```python
>>> import salt.utils.odict
>>> isinstance(salt.utils.odict.OrderedDict(), dict)
True
```

I'm also confused by your recent commit. I thought you wanted pillar to be an OrderedDict, why did you suddenly change this?